formatters/xccov/xccov.go

```diff
@@ -0,0 +1,88 @@
+package xccov
+
+import (
+	"encoding/json"
+	"os"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/codeclimate/test-reporter/env"
+	"github.com/codeclimate/test-reporter/formatters"
+	"github.com/pkg/errors"
+)
+
+var searchPaths = []string{"coverage.json"}
+
+type Formatter struct {
+	Path string
+}
+
+func (f *Formatter) Search(paths ...string) (string, error) {
+	paths = append(paths, searchPaths...)
+	for _, p := range paths {
+		logrus.Debugf("checking search path %s for xccov formatter", p)
+		if _, err := os.Stat(p); err == nil {
+			f.Path = p
+			return p, nil
+		}
+	}
+
+	return "", errors.WithStack(errors.Errorf("could not find any files in search paths for xccov. search paths were: %s", strings.Join(paths, ", ")))
+}
+
+func (r Formatter) Format() (formatters.Report, error) {
+	report, err := formatters.NewReport()
+	if err != nil {
+		return report, err
+	}
+
+	inputXccovFile, err := os.Open(r.Path)
+	if err != nil {
+		return report, errors.WithStack(errors.Errorf("could not open coverage file %s", r.Path))
+	}
+
+	covFile := &xccovFile{}
+	err = json.NewDecoder(inputXccovFile).Decode(&covFile)
+	if err != nil {
+		return report, errors.WithStack(err)
+	}
+
+	gitHead, _ := env.GetHead()
+	for _, target := range covFile.Targets {
+		for _, jsonFile := range target.Files {
+			num := 1
+			sourceFile, err := formatters.NewSourceFile(jsonFile.Path, gitHead)
+			if err != nil {
+				logrus.Warnf("Couldn't find file for path \"%s\" from %s coverage data. Ignore if the path doesn't correspond to an existent file in your repo.", jsonFile.Path, r.Path)
+				continue
```

Comment:
One thing I'm confused by: we have a test here, which is passing, but if I'm reading these lines right, we skip over non-existent files. But the test doesn't supply those files, only the report.

Edit: oh... I see... we overwrite the function that looks up the git blob, in the test... Neat